File: core/src/main/java/net/dontdrinkandroot/wicket/behavior/CssClassAppender.java
/*
 * Copyright (C) 2012-2017 Philip Washington Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.behavior;

import net.dontdrinkandroot.wicket.css.CssClass;
import net.dontdrinkandroot.wicket.model.CssClassClassStringModel;
import net.dontdrinkandroot.wicket.model.StringModel;
import org.apache.wicket.behavior.AttributeAppender;
import org.apache.wicket.model.IModel;

/**
 * Appends a <tt>class</tt> attribute to an element. Can be a String, A {@link CssClass} or a model of a {@link CssClass}.
 *
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class CssClassAppender extends AttributeAppender
{
    public CssClassAppender(final String classToAdd)
    {
        super("class", new StringModel(classToAdd), " ");
    }

    public CssClassAppender(CssClass cssClass)
    {
        super("class", new StringModel(cssClass.getClassString()), " ");
    }

    public CssClassAppender(IModel<? extends CssClass> cssClassModel)
    {
        super("class", new CssClassClassStringModel(cssClassModel), " ");
    }
}

File: core/src/main/kotlin/net/dontdrinkandroot/wicket/behavior/CssClassAppender.kt
package net.dontdrinkandroot.wicket.behavior

import net.dontdrinkandroot.wicket.css.CssClass
import net.dontdrinkandroot.wicket.model.CssClassClassStringModel
import org.apache.wicket.MarkupContainer
import org.apache.wicket.behavior.AttributeAppender
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model

/**
 * Appends a <tt>class</tt> attribute to an element. Can be a String, A [CssClass] or a model of a [CssClass].
 */
open class CssClassAppender : AttributeAppender {

    constructor(classToAdd: String?) : super("class", Model(classToAdd), " ")
    constructor(cssClass: CssClass) : super("class", Model(cssClass.classString), " ")
    constructor(cssClassModel: IModel<out CssClass?>) : super("class", CssClassClassStringModel(cssClassModel), " ")
}

fun cssClass(cssClassModel: IModel<out CssClass?>) = CssClassAppender(cssClassModel)

fun cssClass(cssClass: CssClass) = CssClassAppender(cssClass)

fun cssClass(cssClass: String) = CssClassAppender(cssClass)

fun MarkupContainer.addCssClass(cssClass: CssClass) {
    add(CssClassAppender(cssClass))
}

fun MarkupContainer.addCssClass(cssClass: String) {
    add(CssClassAppender(cssClass))
}

fun cssClasses(vararg cssClasses: CssClass) = CompositeBehavior(cssClasses.map { CssClassAppender(it) }.toMutableList())